x-pack/plugin/monitoring/src/main/java/org/elasticsearch/xpack/monitoring/action/TransportMonitoringMigrateAlertsAction.java

```diff
@@ -0,0 +1,226 @@
+import org.elasticsearch.xpack.monitoring.exporter.Exporters;
+import org.elasticsearch.xpack.monitoring.exporter.MonitoringMigrationCoordinator;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.Collectors;
+
+public class TransportMonitoringMigrateAlertsAction extends TransportMasterNodeAction<MonitoringMigrateAlertsRequest,
+    MonitoringMigrateAlertsResponse> {
+
+    private static final Logger logger = LogManager.getLogger(TransportMonitoringMigrateAlertsAction.class);
+
+    private static final TimeValue ALERT_MIGRATION_INSTALLATION_WAIT_TIMEOUT = new TimeValue(10, TimeUnit.SECONDS);
+    private static final TimeValue ALERT_MIGRATION_BACK_OFF_RETRY = new TimeValue(5, TimeUnit.SECONDS);
+    private static final int ALERT_MIGRATION_MAX_RETRY = 3;
+
+    private final Client client;
+    private final MonitoringMigrationCoordinator migrationCoordinator;
+    private final Exporters exporters;
+
+    @Inject
+    public TransportMonitoringMigrateAlertsAction(Client client, Exporters exporters, MonitoringMigrationCoordinator migrationCoordinator,
+                                                  TransportService transportService, ClusterService clusterService, ThreadPool threadPool,
+                                                  ActionFilters actionFilters, IndexNameExpressionResolver indexNameExpressionResolver) {
+        super(MonitoringMigrateAlertsAction.NAME, transportService, clusterService, threadPool, actionFilters,
+            MonitoringMigrateAlertsRequest::new, indexNameExpressionResolver, MonitoringMigrateAlertsResponse::new,
+            ThreadPool.Names.MANAGEMENT);
+        this.client = client;
+        this.migrationCoordinator = migrationCoordinator;
+        this.exporters = exporters;
+    }
+
+    @Override
+    protected void masterOperation(Task task, MonitoringMigrateAlertsRequest request, ClusterState state,
+                                   ActionListener<MonitoringMigrateAlertsResponse> listener) throws Exception {
+        // First, enable the migration coordinator block
+        if (migrationCoordinator.tryBlockInstallationTasks(ALERT_MIGRATION_INSTALLATION_WAIT_TIMEOUT) == false) {
+            throw new EsRejectedExecutionException("Could not migrate cluster alerts. Timed out waiting to block resource operations.");
+        }
+        // Wrap the listener to unblock resource installation before completing
+        listener = ActionListener.runBefore(listener, migrationCoordinator::unblockInstallationTasks);
+        try {
+            Settings.Builder decommissionAlertSetting = Settings.builder().put(Monitoring.MIGRATION_DECOMMISSION_ALERTS.getKey(), true);
+            client.admin().cluster().prepareUpdateSettings().setPersistentSettings(decommissionAlertSetting)
+                .execute(afterSettingUpdate(listener));
```

Comment:
What thread is all this work happening on ?

Reply:
Hmm, if the master node it seems that this runs on a master service thread that completed the settings update. If not, I believe it's a network thread. I'll fork it over to management after the request succeeds.